pkg/kubelet/kubelet.go

```diff
@@ -3843,3 +3843,36 @@ func (kl *Kubelet) shapingEnabled() bool {
 func (kl *Kubelet) GetNodeConfig() cm.NodeConfig {
 	return kl.containerManager.GetNodeConfig()
 }
+
+// killPodNow kills a pod.
+// The pod status is updated, and then it is killed with the specified grace period.
+// This function must block until either the pod is killed or an error is encountered.
+// Arguments:
+// pod - the pod to kill
+// status - the desired status to associate with the pod (i.e. why its killed)
+// gracePeriodOverride - the grace period override to use instead of what is on the pod spec
+func (kl *Kubelet) killPodNow(pod *api.Pod, status api.PodStatus, gracePeriodOverride *int64) error {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	var result error
+	go func() {
+		defer utilruntime.HandleCrash(func(panicReason interface{}) {
+			result = fmt.Errorf("KilLPodNow goroutine panicked: %v", panicReason)
+		})
+		onCompleteFunc := func(err error) {
+			defer wg.Done()
+			result = err
+		}
+		kl.podWorkers.UpdatePod(&UpdatePodOptions{
```

Comment:
Maybe I'm missing something.... How do you prevent the update from being overwritten? AFAIK, the lastUndeliverable can always be overwritten by a new update. The `OnCompleteFunc` may never be called once the update gets overwritten.

The current design of pod workers does not execute every update. It cares only about the latest pod spec/status. In your case, you may want to do smart "update merging" in PodWorkers or add a special state in PodWorkers to prevent the worker from receiving anymore updates.